tokio-util/tests/task_join_queue.rs

```diff
@@ -0,0 +1,228 @@
+#![warn(rust_2018_idioms)]
+
+use tokio::sync::oneshot;
+use tokio::task::yield_now;
+use tokio::time::Duration;
+use tokio_test::{assert_pending, assert_ready, task};
+use tokio_util::task::JoinQueue;
+
+#[tokio::test]
+async fn test_join_queue_no_spurious_wakeups() {
+    let (tx, rx) = oneshot::channel::<()>();
+    let mut join_queue = JoinQueue::new();
+    join_queue.spawn(async move {
+        let _ = rx.await;
+        42
+    });
+
+    let mut join_next = task::spawn(join_queue.join_next());
+
+    assert_pending!(join_next.poll());
+
+    assert!(!join_next.is_woken());
+
+    let _ = tx.send(());
+    yield_now().await;
+
+    assert!(join_next.is_woken());
+
+    let output = assert_ready!(join_next.poll());
+    assert_eq!(output.unwrap().unwrap(), 42);
+}
+
+#[tokio::test]
+async fn test_join_queue_abort_on_drop() {
+    let mut queue = JoinQueue::new();
+
+    let mut recvs = Vec::new();
+
+    for _ in 0..16 {
+        let (send, recv) = oneshot::channel::<()>();
+        recvs.push(recv);
+
+        queue.spawn(async move {
+            // This task will never complete on its own.
+            futures::future::pending::<()>().await;
+            drop(send);
+        });
+    }
+
+    drop(queue);
+
+    for recv in recvs {
+        // The task is aborted soon and we will receive an error.
+        assert!(recv.await.is_err());
+    }
+}
+
+#[tokio::test]
+async fn test_join_queue_alternating() {
+    let mut queue = JoinQueue::new();
+
+    assert_eq!(queue.len(), 0);
+    queue.spawn(async {});
+    assert_eq!(queue.len(), 1);
+    queue.spawn(async {});
+    assert_eq!(queue.len(), 2);
+
+    for _ in 0..16 {
+        let res = queue.join_next().await.unwrap();
+        assert!(res.is_ok());
+        assert_eq!(queue.len(), 1);
+        queue.spawn(async {});
+        assert_eq!(queue.len(), 2);
+    }
+}
+
+#[tokio::test(start_paused = true)]
+async fn test_join_queue_abort_all() {
+    let mut queue: JoinQueue<()> = JoinQueue::new();
+
+    for _ in 0..5 {
+        queue.spawn(futures::future::pending());
+    }
+    for _ in 0..5 {
+        queue.spawn(async {
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        });
+    }
+
+    // The join queue will now have 5 pending tasks and 5 ready tasks.
+    tokio::time::sleep(Duration::from_secs(2)).await;
+
+    queue.abort_all();
+    assert_eq!(queue.len(), 10);
+
+    let mut count = 0;
+    while let Some(res) = queue.join_next().await {
+        if count < 5 {
+            assert!(res.is_err());
+            assert!(res.unwrap_err().is_cancelled());
```

Comment:
```suggestion
            assert!(res.unwrap_err().is_cancelled());
```

The `unwrap_err` can cover the `assert!(res.is_err())`.